clippy_lints/src/returns.rs

```diff
@@ -435,3 +437,13 @@ fn extend_span_to_previous_non_ws(cx: &LateContext<'_>, sp: Span) -> Span {
 
     sp
 }
+
+fn is_let_else_block(tcx: TyCtxt<'_>, e: &Expr<'_>) -> bool {
+    for (_, node) in tcx.hir().parent_iter(e.hir_id) {
+        if let Node::Local(Local { els: Some(_), .. }) = node {
+            return true;
+        }
+    }
+
+    false
+}
```

Comment:
Pretty nit picky:

I feel like the `if let Node::Local(Local { els: Some(_), .. })` check could be a bit too general, as it doesn't check, that the expression comes from the else branch:

```rs
fn x() -> Result<(), ()> {
    let Some(_): Option<()> = ({
        return Err(())?;
    }) else {
        panic!();
    };
}
```

Even if that's pretty weird code...

Could you try if this works:

```rs
fn is_let_else_block(tcx: TyCtxt<'_>, e: &Expr<'_>) -> bool {
    let mut child_id = e.hir_id;
    for (parent_id, node) in tcx.hir().parent_iter(e.hir_id) {
        if let Node::Local(Local { els: Some(els), .. }) = node
            && els.hir_id == child_id
        {
            return true;
        }
        child_id = parent_id;
    }

    false
}
```

Then we could probably move this into `clippy_utils` in case other lints need something like this :). If not, it's totally fine. I think this code is already good enough of a heuristic, to avoid FPs and r+ the PR :)

Reply:
🤔 yeah fair enough, if we're going to move this to util, then this needs to be as precise as possible.

```rust
fn x() -> Result<(), ()> {
    let Some(_): Option<()> = ({
        return Err(())?;
    }) else {
        panic!();
    };
}
```

This is kind of outrageous for an FN of this lint, there's more to worry about than a needless return 😅

Reply:
Oh, just remembered that we'd need to ignore this as well

```rust
fn x() -> Result<(), ()> {
    let Some(_): Option<()> = ({
        return Err(())?;
    }) else {
        panic!();
    };
}
```

removing the return here leads to more complications than benefits; e.g. here it leads to compilation error, and if we have the same type, it might change the behavior.

> Could you try if this works:

But this does work